### PR TITLE
[FedEx Integration]

### PR DESCRIPTION
**Pull request by patilsangram**
This is FedEx Integration PR covering,

1. Required Masters/DocType and updates.
2. Re-thought/structure of Shipping Plan as a standalone DocType
3. Association with DN as reference DocType (Currently supported)
4. Provision to Make Shipping Plan from DN.
4. Validations on Shipping Plan.

Please, do review and comment on the same.

Expected next level of updates will include,

1. Revisit field mapping and updates  
2. FedEx related validations
3. Filters as appropriate.
4. Hit FedEx 

References..
[OLD PR](https://github.com/frappe/erpnext/pull/11541)
[Issue](https://github.com/frappe/erpnext/issues/11468)
[Discuss Thread](https://discuss.erpnext.com/t/erpnext-foundation-pune-chapter-code-sprint-3-fedex-integration/32803)

Indictrans.

### CODE REVIEW

**Comment by revant** [comment]
- Can Postal Code be generalised instead of Fedex Postal Code? It can be re-used
- `Package` to `Fedex Package` can be dynamic link instead of Link? DHL Package/UPS package can be set like that?
- Shipper can become linked with supplier? Supplier contacts can be used for multiple persons interacting from shipper party
- Shipping Plan tied with only delivery note?
    - Can it be also tied with Purchase Receipt?
    - Stock Entry -> Material Issue / Receipt can also generate shipping in some cases. Intra warehouse shipping.

**Review by revant — COMMENTED** [review]
Refer Comment

**Comment by patilsangram** [comment]
Thank you @revant. I will do the necessary changes and send it again.

**Comment by saurabh6790** [comment]
@patilsangram Also add some screenshots

**Comment by patilsangram** [comment]
- Can Postal Code be generalised instead of Fedex Postal Code? It can be re-used
 
>> Okay we can rename it as Postal Code

- Package to Fedex Package can be dynamic link instead of Link? DHL Package/UPS package can be set like that?
	
>> If we create Package no as Dynamic field then we have to create Package Type as Link field.

- Shipper can become linked with supplier? Supplier contacts can be used for multiple persons interacting from shipper party
	
>> We can add Supplier link on shipper

- Shipping Plan tied with only delivery note? Can it be also tied with Purchase Receipt?
Stock Entry -> Material Issue / Receipt can also generate shipping in some cases. Intra company warehouse shipping.
	
>> Yes, it is in plan.Currently we are mapped it from Delivery Note. In future we can provide Make option on above mentioned form to Make Shipping Plan from it.


- Edit: Also I Feel, Company needs child table for various different numbers like TIN, PAN, etc. because we need to keep adding removing such numbers for domains, integrations, localisations, etc.

>> Let's wait for more comments from community on this.

**Comment by revant** [comment]
> If we create Package no as Dynamic field then we have to create Package Type as Link field.

For this PR we can go with Link and have a complete integration with FedEx first.

> Let's wait for more comments from community on this.

:+1: We'll wait for comments.

**Comment by patilsangram** [comment]
Hi,

As per suggestions, I have renamed FedEx Postal Code to Postal Code. 
Also, added supplier link on Shipper Master. Let me know if anything. So, we can move ahead.

**Comment by patilsangram** [comment]
@saurabh6790 

These are the some Master's screenshot.

- FedEx Settings

image

- Shipper

image

- Shipping Plan

image

Shipping Plan Item [Table] - Item details
image

Package [Table] - for Package details
image

Package Item [Table] - for item distribution in Packges
image

[ Other details ]
image

- Postal Code

image

**Comment by joshreeder** [comment]
What needs to happen on this?

**Comment by patilsangram** [comment]
We would require the FedEx credentials to validate the API's. We expired the earlier one available and not sure how much time will it take to get one fresh. If anybody has an access can spare for testing would be really great to expedite on hold implementation. We are sure that will not charge you anything since we can cancel a planned pick-up (or generally not charged unless pick-up from your end).

File: erpnext/logistics/controller/fedex_controller.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from fedex_config import fedex_config
+from frappe.utils import get_datetime
+from fedex.services.pickup_service import FedexCreatePickupRequest
+
+class FedExController():
+	""" A Higher-Level wrapper for Fedex python library
+		which handles API like Shipment, Tracking, GET Rate
+		& other supplementary tasks. """
+
+	uom_mapper = {"Kg":"KG", "LB":"LB"}
```
Comment (rohitwaghchaure) [inline erpnext/logistics/controller/fedex_controller.py]: Why uom mapper?

**Pull request closed without merging**